File: calcul_moyennes_30min.R
# Script R pour calculer des moyennes mobiles sur 30 observations
# à partir d'un fichier CSV de données de rayonnement.

# Charger les bibliothèques nécessaires
library(readr)
library(dplyr)
library(zoo)

# Chemin du fichier d'entrée
fichier_entree <- "/DATA/BILOS/RAW/ICOS/ANNEEENCOURS/icos/rayonnement/output/ICOS_RAY_20S.dat_all.csv"

# Chemin du fichier de sortie
fichier_sortie <- "/DATA/BILOS/RAW/ICOS/ANNEEENCOURS/icos/rayonnement/output/ICOS_RAY_20S.dat_all_30min_mean.csv"

# Lire le fichier CSV
# Supposons que la première colonne est le timestamp et les autres sont numériques
donnees <- read_csv2(
  fichier_entree,
  col_names = TRUE,
  na = c("", "NA", "NaN"),
  locale = readr::locale(encoding = "UTF-8")
)

# Vérifier que le fichier a été lu correctement
if (ncol(donnees) == 0) {
  stop("Erreur : Le fichier est vide ou n'a pas été lu correctement.")
}

# Afficher les premières lignes pour vérifier
head(donnees)

# Identifier les colonnes numériques (en excluant la première colonne, supposée être le timestamp)
colonnes_numeriques <- sapply(donnees, is.numeric)
noms_colonnes_numeriques <- names(donnees)[colonnes_numeriques]

# Si la première colonne n'est pas numérique, la considérer comme le timestamp
if (!colonnes_numeriques[1]) {
  timestamp_col <- 1
  colonnes_numeriques[1] <- FALSE
  noms_colonnes_numeriques <- names(donnees)[colonnes_numeriques]
} else {
  timestamp_col <- NULL
}

# Calculer les moyennes mobiles sur 30 observations pour chaque colonne numérique
for (col in noms_colonnes_numeriques) {
  # Utiliser rollmean de zoo pour calculer la moyenne mobile
  # align = "right" pour que la moyenne soit alignée avec la dernière observation de la fenêtre
  # fill = NA pour remplir les premières valeurs avec NA
  donnees[[paste0(col, "_30min_mean")]] <- rollmean(
    donnees[[col]],
    k = 30,
    fill = NA,
    align = "right"
  )
}

# Sauvegarder le résultat dans un nouveau fichier CSV
write_csv2(
  donnees,
  fichier_sortie,
  na = "",
  row_names = FALSE
)

# Afficher un message de confirmation
cat("Calcul des moyennes mobiles sur 30 observations terminé.\n")
cat("Fichier de sortie :", fichier_sortie, "\n")

File: calcul_moyennes_30min.py
import pandas as pd
import os


def calculer_moyennes_mobiles_30(
    fichier_entree: str,
    fichier_sortie: str,
    colonne_timestamp: str = None,
) -> None:
    """
    Calcule des moyennes mobiles sur 30 observations pour chaque colonne numérique d'un fichier CSV.

    Args:
        fichier_entree: Chemin du fichier CSV d'entrée.
        fichier_sortie: Chemin du fichier CSV de sortie.
        colonne_timestamp: Nom de la colonne de timestamp (optionnel, pour éviter de calculer la moyenne sur cette colonne).
    """
    # Lire le fichier CSV
    df = pd.read_csv(fichier_entree, encoding="utf-8", na_values=["", "NA", "NaN"])

    if df.empty:
        raise ValueError("Le fichier est vide ou n'a pas été lu correctement.")

    print("Premières lignes du fichier :")
    print(df.head())

    # Identifier les colonnes numériques
    colonnes_numeriques = df.select_dtypes(include=["float64", "int64"]).columns.tolist()

    # Si une colonne de timestamp est spécifiée, l'exclure des colonnes numériques
    if colonne_timestamp and colonne_timestamp in colonnes_numeriques:
        colonnes_numeriques.remove(colonne_timestamp)

    # Calculer les moyennes mobiles sur 30 observations pour chaque colonne numérique
    for col in colonnes_numeriques:
        nouvelle_colonne = f"{col}_30min_mean"
        df[nouvelle_colonne] = df[col].rolling(window=30, min_periods=1).mean()

    # Sauvegarder le résultat dans un nouveau fichier CSV
    df.to_csv(fichier_sortie, index=False, encoding="utf-8", na_rep="")

    print(f"\nCalcul des moyennes mobiles sur 30 observations terminé.")
    print(f"Fichier de sortie : {fichier_sortie}")


if __name__ == "__main__":
    # Chemins des fichiers
    FICHIER_ENTREE = "/DATA/BILOS/RAW/ICOS/ANNEEENCOURS/icos/rayonnement/output/ICOS_RAY_20S.dat_all.csv"
    FICHIER_SORTIE = "/DATA/BILOS/RAW/ICOS/ANNEEENCOURS/icos/rayonnement/output/ICOS_RAY_20S.dat_all_30min_mean.csv"

    # Exécuter la fonction
    calculer_moyennes_mobiles_30(
        fichier_entree=FICHIER_ENTREE,
        fichier_sortie=FICHIER_SORTIE,
        colonne_timestamp="TIMESTAMP",  # Remplacez par le nom de votre colonne de timestamp si nécessaire
    )
